packages/vite-plugin-rsc/src/plugin.ts

```diff
@@ -0,0 +1,212 @@
+import { existsSync, readFileSync } from "node:fs";
+interface ASTNode {
+  type: string;
+  start: number;
+  end: number;
+  body?: Array<ASTNode>;
+  id?: ASTNode;
+  expression?: ASTNode;
+  declaration?: ASTNode;
+  declarations?: Array<ASTNode>;
+  name?: string;
+  specifiers?: Array<ASTNode>;
+
+  value?: string;
+  exported?: ASTNode;
+}
+
+/**
+ * Checks if the node has a string literal at the top level that matches the statement
+ */
+const hasPragma = (ast: ASTNode, statement: string) =>
+  ast.body?.some((node) => {
+    return (
+      node.type === "ExpressionStatement" &&
+      node.expression?.type === "Literal" &&
+      node.expression.value === statement
+    );
+  });
+
+/**
+ * Finds all the named and default exports of a module
+ */
+
+const getExports = (ast: ASTNode) => {
+  const exports: Array<string> = [];
+  ast.body?.forEach((node) => {
+    if (node.type === "ExportDefaultDeclaration") {
+      exports.push("default");
+    }
+    if (node.type === "ExportNamedDeclaration") {
+      if (node.declaration?.type === "VariableDeclaration") {
+        node.declaration?.declarations?.forEach((declaration) => {
+          const name = declaration?.id?.name;
+          if (name) {
+            exports.push(name);
+          }
+        });
+        return;
+      }
+
+      if (node.declaration?.type === "FunctionDeclaration") {
+        const name = node.declaration?.id?.name;
+        if (name) {
+          exports.push(name);
+        }
+        return;
+      }
+
+      if (node.specifiers?.length) {
+        node.specifiers.forEach((specifier) => {
+          const name = specifier?.exported?.name;
+          if (name) {
+            exports.push(name);
+          }
+        });
+      }
+    }
+  });
+  return exports;
+};
+
+export default function plugin({
+  serverDist = "dist/server",
+  clientDist = "dist/static",
+}: {
+  serverDist?: string;
+  clientDist?: string;
+}): Plugin {
+  const clientPragma = "use client";
+  // const serverPragma = "use server";
+  let externals = new Set<string>();
+  let config: ResolvedConfig;
+  let isSsr: boolean;
+  let isBuild: boolean;
+  let manifest: Manifest = {};
+
+  const bundleMap = new Map();
+  const clientModuleId = "virtual:client-bundle-map";
+  const resolvedClientModuleId = "\0" + clientModuleId;
+  const serverModuleId = "virtual:server-bundle-map";
+  const resolvedServerModuleId = "\0" + serverModuleId;
+
+  const clientBundleMapFilename = "client-bundle-map.json";
+  const serverBundleMapFilename = "server-bundle-map.json";
+
+  return {
+    name: "vite-plugin-extract-server-components",
+
+    config(config) {
+      config.build ||= {};
+      if (config.build.ssr) {
+        const manifestPath = path.join(
+          config.root || "",
+          clientDist,
+          "manifest.json"
+        );
+        if (existsSync(manifestPath)) {
+          manifest = JSON.parse(readFileSync(manifestPath, "utf-8"));
+        }
```

Comment:
The manifest.json is generated when building the _client_ code, and maps module names to chunk filenames. e.g.

```json
{
  "index.html": {
    "file": "assets/index-3cfb730f.js",
    "isEntry": true,
    "src": "index.html"
  },
  "src/assets/impala.png": {
    "file": "assets/impala-0936a6d6.png",
    "src": "src/assets/impala.png"
  },
  "src/components/button.tsx": {
    "file": "assets/button-8f2948fa.js",
    "isEntry": true,
    "src": "src/components/button.tsx"
  }
}
```

During SSR we load the manifest so we can use it to find client chunk names